File: TreeUtils.java
import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class TreeUtils {
    private static TreeNode toNode(final String str) {
        if (str.length() == 0 || str.toLowerCase().equals("null")) {
            return null;
        }
        return new TreeNode(Integer.valueOf(str));
    }

    public static TreeNode buildTree(String[] nums) {
        if (nums.length == 0 || toNode(nums[0]) == null)
            return null;
        TreeNode root = toNode(nums[0]);

        Queue<TreeNode> qu = new LinkedList<>();
        qu.add(root);
        for (int i = 1; i < nums.length && !qu.isEmpty();) {
            TreeNode node = qu.poll();
            TreeNode next = toNode(nums[i]);
            if (next != null) {
                node.left = next;
                qu.add(next);
            }
            i++;
            if (i < nums.length) {
                next = toNode(nums[i]);
                if (next != null) {
                    node.right = next;
                    qu.add(next);
                }
            }
            i++;
        }

        return root;
    }

    public static void printTree(TreeNode root) {
        if (root == null)
            return;
        Queue<TreeNode> qu = new LinkedList<>();
        qu.add(root);
        boolean finished = false;
        while (!qu.isEmpty() && !finished) {
            finished = true;
            int levelNum = qu.size();
            for (int i = 0; i < levelNum; i++) {
                TreeNode node = qu.poll();
                if (node != null) {
                    finished = false;
                    System.out.printf("%d ", node.val);
                    qu.add(node.left);
                    qu.add(node.right);
                } else {
                    System.out.printf("*");
                    qu.add(null);
                    qu.add(null);
                }
            }
            System.out.printf("\n");
        }
    }

    public static TreeNode getNode(TreeNode root, int val) {
        if (root == null)
            return null;
        if (root.val == val)
            return root;
        TreeNode left = getNode(root.left, val), right = getNode(root.right, val);
        if (left != null)
            return left;
        return right;
    }
}

File: TreeUtils.kt
import java.util.*

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class TreeUtils {
    companion object {
        fun buildTree(nums: Array<Int?>): TreeNode? {
            if (nums.isEmpty())
                return null
            val root = TreeNode(nums[0]!!)
            val qu = LinkedList<TreeNode>()
            qu.add(root)
            var i = 1
            while (i < nums.size) {
                if (qu.isEmpty()) {
                    break
                }
                val node = qu.poll()!!
                nums[i]?.let {
                    TreeNode(it)
                }?.let {
                    node.left = it
                    qu.add(it)
                }
                i++
                if (i < nums.size) {
                    nums[i]?.let {
                        TreeNode(it)
                    }?.let {
                        node.right = it
                        qu.add(it)
                    }
                }
                i++
            }

            return root
        }

        fun printTree(root: TreeNode?) {
            if (root == null)
                return
            val qu = LinkedList<TreeNode?>()
            qu.add(root)
            var finished = false
            while (!qu.isEmpty() && !finished) {
                finished = true
                val levelNum = qu.size
                for (i in 0 until levelNum) {
                    val node = qu.poll()
                    if (node != null) {
                        finished = false
                        print("%d ".format(node.`val`))
                        qu.add(node.left)
                        qu.add(node.right)
                    } else {
                        System.out.printf("*")
                        qu.add(null)
                        qu.add(null)
                    }
                }
                println()
            }
        }

        fun getNode(
            root: TreeNode?,
            `val`: Int
        ): TreeNode? {
            if (root == null)
                return null
            if (root.`val` == `val`)
                return root
            val left = getNode(root.left, `val`)
            val right = getNode(root.right, `val`)
            if (left != null)
                return left
            return right
        }
    }
}